### DISCUSSION TITLE
zsh: permission denied

### DISCUSSION THREAD

**Discussion by cool-delete** [ohmyzsh/ohmyzsh#10052] [Help]
Hello, I am a newbie, I don't know where to solve this problem, I can only come here, thank you.

![image](https://user-images.githubusercontent.com/13638610/125439306-78a863c5-8f9f-4abe-83a6-843d7049981d.png)

Json-service is a global installed npm package when I use BASH without any problems but is prompted in Zsh.

`zsh: permission denied`

Not all bags are like this, but npm is normal.

**Comment by mcornella** [comment]
Did you solve this? If not, what do `which -a json-server` and `ls -l $(which json-server)` show?

**Reply by cool-delete** [reply]
![image](https://user-images.githubusercontent.com/13638610/133957028-c1e67774-4f72-4f09-b3d3-d8b8158fc134.png)

**Reply by cool-delete** [reply]
During this period, I can't count the number of updates of ohmyzsh, and it has not been resolved until now. I need to switch to bash to use the command

**Reply by rwmitchell** [reply]
What does those commands show in bash, where it does work?

**Reply by cool-delete** [reply]
The screenshot mentioned above shows the result of the command. The path of bash and zsh is the same

**Reply by mcornella** [reply]
When you run `bash`, which you say `json-server` works there, if you run the commands above *inside bash* what do you get?

**Comment by rwmitchell** [comment]
Have you considered you may have modified your bash path to include json-server that you did not add to your zsh path?

**Reply by cool-delete** [reply]
Sorry, are you talking about the content of bash.rc? I don’t understand what you mean

**Reply by rwmitchell** [reply]
Either the path to ```json-server``` is specified in your bash setup (more likely ```.bashrc```, not bash.rc) or it is defined as a function or alias there. 

It is very unlikely OMZ is having anything to do with it this particular command not being found.

**Reply by cool-delete** [reply]
umm ... Still can't understand. This is the environment variable of my .bashrc
```
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
export NODE_PATH="/usr/local/lib/node_modules"
export NODE_PATH="/usr/local/lib/node_modules"
export NODE_PATHG="/root/.nvm/versions/node/v12.18.2/lib/node_modules"
export NODE_bin="~/.nvm/versions/node/v12.18.2/bin"
export PATH=$PATH:$NODE_PATH:$NODE_PATHG:$NODE_bin
```

.zshrc
``` 
export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
#export NVM_DIR="$([ -z "${XDG_CONFIG_HOME-}" ] && printf %s "${HOME}/.nvm" || printf %s "${XDG_CONFIG_HOME}/nvm")"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh" 
export NODE_PATH="/usr/local/lib/node_modules"
export NODE_PATHG="/root/.nvm/versions/node/v12.18.2/lib/node_modules"
export NODE_bin="~/.nvm/versions/node/v12.18.2/bin"
export NODE_bin14="~/.nvm/versions/node/v14.4.0/bin"
export PATH=$PATH:$NODE_PATH:$NODE_PATHG:$NODE_bin:$NODE_bin14:$NVM_DIR
# source ~/.bash_profile
source $ZSH/oh-my-zsh.sh
source "/code/emsdk/emsdk_env.sh"
```

**Reply by rwmitchell** [reply]
Just tell us what ```which -a json_server``` and ```ls -l $(which json-server```
reports WHEN it works from BASH!

**Reply by cool-delete** [reply]
in bash 
`which -a json_server` no value is returned
![image](https://user-images.githubusercontent.com/13638610/135476591-ed3d0670-9125-490c-8574-1daa6a7bc9b7.png)
![image](https://user-images.githubusercontent.com/13638610/135476767-bf5efb17-80aa-44b0-b168-42b4b6686597.png)

**Reply by rwmitchell** [reply]
If the first command didn't show anything, second command isn't going to work either.
where is proof json-server works?
why are you running as root when you clearly don't know what you're doing?

**Reply by cool-delete** [reply]
The windows sub system is used, and root is the default account..
Do you mean that it will be normal if I am not under the root account, but can my work under the root account be migrated to other sub-accounts?

**Reply by rwmitchell** [reply]
Running as root is a poor practice, regardless of knowledge level.  root should only be necessary in select cases and for that there is sudo.    Given you're learning from MS, I'm not surprised.

Can you demonstrate json-server works in bash?  If it does, then ```which json-server``` or ```which -a json-server``` should show where it exists.  Once we know that, we can then figure out how to get it to work in zsh. 

However, this assumes a sane OS, which I'm not sure anything Windows qualifies.

You should be able to run ```json-server --version``` to demonstrate it works.
After you demonstrate it works, then show again ```which -a json-server```

**Reply by cool-delete** [reply]
Oh no..

`json-server --version`

The return value is 0.16.3

bash operation result
![image](https://user-images.githubusercontent.com/13638610/135581463-34ab540b-8662-4e41-b525-f362780e7725.png)
Is no return because my npm uses nvm management?
`npm list -g --depth 0`

![image](https://user-images.githubusercontent.com/13638610/135582452-5ad7183d-06af-4f54-a687-6c9415a23ffc.png)

**Reply by mcornella** [reply]
When I install `json-server` via `npm install -g json-server`, it puts it in `~/.npm/bin`, so that should be in `$PATH`. You can check where npm installs globals with `npm root -g`: if it's `$HOME/.npm/lib/node_modules`, then the global bin folder will be `~/.npm/bin`. If it's `/usr/local/lib/node_modules`, the global bin folder will be probably `/usr/local/bin`. Either of those need to be in `$PATH` when running from zsh, otherwise you will get the command not found error.

**Reply by cool-delete** [reply]
@mcornella  thx. I use the nvm tool to manage the node version. When I use `npm root -g` to view the path
![image](https://user-images.githubusercontent.com/13638610/136101187-5c006763-a7d4-4df4-95f8-ef26279ba244.png)

![image](https://user-images.githubusercontent.com/13638610/136101136-ea7809bb-144c-462c-865c-b8f8b95bfc7d.png)
